Architecture deep dive (#167)
cambiatus/backend: .github/deep-dive.md

# Table of Contents
- **[Architecture Overview](#Architecture-Overview)**
- **[Flow of cambiatus data](#Flow-of-cambiatus-data)**
  - **[Creating a new community](#Creating-a-new-community)**
  - **[Updating community](#Updating-community)**
- **[Resources](#Resources)**

# Architecture Overview

> Cambiatus is a DApp (decentralized application) that leverages [EOS](https://training.eos.io/) blockchain protocol, here are few resource to checkout if you are new to DApp, [[1]](https://www.freecodecamp.org/news/what-is-a-dapp-a-guide-to-ethereum-dapps/), [[2]](https://medium.com/proof-systems/a-dapp-is-not-a-protocol-824411a55582).

Reviewer: ```suggestion
Cambiatus is a DApp (decentralized application) that leverages the [EOS](https://training.eos.io/) blockchain protocol. Here are a few resources to checkout if you are new to DApp: [what is a DApp](https://www.freecodecamp.org/news/what-is-a-dapp-a-guide-to-ethereum-dapps/), [A DApp is not a protocol](https://medium.com/proof-systems/a-dapp-is-not-a-protocol-824411a55582).
```

Besides some typos, I think giving meaningful information about the links before people click on them is usually better than just giving them an index

Author: Agreed and updated.

> Even though blockchain brings many benefits to an applicaiton it also has its limitations. The major limitation of blockchain is the ability to query and fetch data in a timely manner. In order to over come this limitation we took a hybrid approach: we store all data on the blockchain and replicate partial data on a traditional database. With this approach we gain the benefits of blockchain - security and immutability of our data; efficient and user friendly retrival of our data from a traditional database.

Reviewer: ```suggestion
Even though blockchain brings many benefits to an applicaiton it also has its limitations. The major limitation of blockchain is the ability to query and fetch data in a timely manner. In order to over come this limitation we took a hybrid approach: we store all data on the blockchain and replicate partial data on a traditional database. With this approach we gain the benefits of blockchain - security and immutability of our data - and the benefits of traditional databases - efficient and user friendly retrieval of data.
```

# Flow of cambiatus data 
In this section we'll explore how our data flows through our application. Here are the objectives for this section:

* Understand the interaction between frontend and blockchain
* Transferring of data from blockchain to database
* Retrival and displaying of the data

When learning a new concept it's always great to use aexamples, so we'll walkthrough two use cases:
- **[Creating a new community](#Creating-a-new-community)**
- **[Updating community](#Updating-community)**

## Creating a new community
Our [frontend](https://github.com/cambiatus/frontend) is a static application built using Elm; the frontend sends *transactions* and triggers *actions* on the blockchain using [eosjs](https://github.com/EOSIO/eosjs) library. 

#### Difference between action and transaction?
Here is a great explaintation of what action and transcation mean in context of EOS:
> "An action is a unit of code to be executed inside a transaction. A transaction is a set of one or more actions which will execute or fail completely." [source](https://forum.ivanontech.com/t/reading-assignment-eos-basics/3085/6)

### Sending action

> Let's say we want to create a new community called `0,TST`. In order to do so we need to do the following steps:

Reviewer: ```suggestion
Let's say we want to create a new community identified by the symbol `0,TST`. In order to do so we need to do the following steps:
```

Communities also have a `name` field, so saying it's called by the symbol might be confusing

1. Compose a transaction for the `create` community `action`
2. Sign and send the transaction to the blockchain

A transaction has the following structure
```
{
  account: "cambiatus.cm",
  name: "create",
  authorization: [{ actor: "janedoe", permission: "active" }],
  data: {

>     cmm_asset: 0 TST
>     creator: janedoe12334
>     name: TESTCOM
>     description: A test community

Reviewer: ```suggestion
    cmm_asset: "0 TST",
    creator: "janedoe12334",
    name: "TESTCOM",
    description: "A test community"
```

The new community has been added to the blockchain! However, we sill need to send the data to our database so our frontend could fetch and display it to the user. We can accompolish this by leveraging an event driven library developed by EOS called [demux-js](https://github.com/EOSIO/demux-js-eos).

#### How demux-js works
1. The `Action Reader` listens for any changes in blockchain. When it detects a change it will pass the data to the `Action Watcher`.
2. `Action Watcher` triggers the `Action Handler`, which in turn updates our database with the data retrieved from the blockchain.

### Updating frontend with the new data
Once an action has been pushed to the blockchain, we *optimistically assume* that data has trickled down to our database. The current method is very unreliable for querying for the data; however, we are looking to leverage Graphql subscription for this [see here](https://github.com/cambiatus/backend/issues/148).

## Updating community
The process for updating a community is similar to **[creating a new community](#Creating-a-new-community)** the only difference is how we react once we have pushed the transaction to the blockchain.

> After we push a transaction, on a successful operation we recieve a `transactionId` this indicated that our data got added to the blockchain. 
> 
> Or our transacction can fail. It can fail because of two reasons:

Reviewer: ```suggestion
After we push a transaction, on a successful operation we receive a `transactionId`, which indicates that our data has been added to the blockchain. 

Or our transaction can fail. It can fail for two reasons:
```
